src/components/application_manager/test/rpc_passing_handler_test.cc

```diff
@@ -0,0 +1,429 @@
+  EXPECT_CALL(mock_rpc_service_,
+              ManageMobileCommand(Pointee(forwarded_request),
+                                  am::commands::Command::SOURCE_MOBILE));
+
+  bool result = rpc_passing_handler_->RPCPassThrough(unsupported_response);
+  EXPECT_EQ(result, true);
+}
+
+TEST_F(RPCPassingHandlerTest,
+       RPCPassingTest_UNSUPPORTED_REQUEST_ForwardToMobile) {
+  smart_objects::SmartObject unsupported_response = CreatePassThroughResponse(
+      kConnectionKey_NAV_ASP, kCorrelationId, "UNSUPPORTED_REQUEST", false);
+  smart_objects::SmartObject forwarded_request =
+      CreatePassThroughRequest(kConnectionKey_MEDIA_ASP, kCorrelationId);
+
+  app_services_.push_back(CreateAppService(
+      kConnectionKey_NAV_ASP, "Navigation service", "NAVIGATION"));
+  app_services_.push_back(
+      CreateAppService(kConnectionKey_MEDIA_ASP, "Media service", "MEDIA"));
+
+  PassthroughParams request_params{
+      .connection_key = kConnectionKey_ASC,
+      .correlation_id = kCorrelationId,
+      .message = CreatePassThroughRequest(kConnectionKey_ASC, kCorrelationId),
+      .forwarded_message =
+          CreatePassThroughRequest(kConnectionKey_NAV_ASP, kCorrelationId),
+  };
+
+  SendRequestToASP(request_params);
+
+  // Call RPCPassThrough with response smart object
+  // Will cycle to next compatible active app service
+  EXPECT_CALL(mock_app_manager_settings_, rpc_pass_through_timeout());
+  EXPECT_CALL(mock_app_manager_, GetRPCService());
+  EXPECT_CALL(mock_rpc_service_,
+              SendMessageToMobile(Pointee(forwarded_request), false));
+
+  bool result = rpc_passing_handler_->RPCPassThrough(unsupported_response);
+  EXPECT_EQ(result, true);
+}
+
+TEST_F(RPCPassingHandlerTest, RPCPassingTest_REQUEST_Timeout) {
+  uint32_t timeout_in_ms = 4;
+  std::shared_ptr<TestAsyncWaiter> waiter = std::make_shared<TestAsyncWaiter>();
+
+  app_services_.push_back(CreateAppService(
+      kConnectionKey_NAV_ASP, "Navigation service", "NAVIGATION"));
+
+  PassthroughParams request_params{
+      .connection_key = kConnectionKey_ASC,
+      .correlation_id = kCorrelationId,
+      .message = CreatePassThroughRequest(kConnectionKey_ASC, kCorrelationId),
+      .forwarded_message =
+          CreatePassThroughRequest(kConnectionKey_NAV_ASP, kCorrelationId),
+  };
+
+  SendRequestToASP(request_params, timeout_in_ms);
+
+  // Request timeout will trigger perform next request
+  // Will cycle to core (no other app services in list)
+  EXPECT_CALL(mock_app_manager_, GetRPCHandler());
+  EXPECT_CALL(mock_rpc_handler_,
+              ValidateRpcSO(request_params.message, _, _, false))
+      .WillOnce(Return(true));
+  EXPECT_CALL(mock_app_manager_, GetRPCService());
+  EXPECT_CALL(mock_rpc_service_,
+              ManageMobileCommand(Pointee(request_params.message),
+                                  am::commands::Command::SOURCE_MOBILE))
+      .WillOnce(DoAll(NotifyTestAsyncWaiter(waiter), Return(true)));
```

Comment:
Technically, these `EXPECT_CALL`s should be above `SendRequestToASL`

Reply:
I guess but these calls aren't directly triggered during the execution of SendRequestToASP(or the RPCPassthrough function) but rather after the timeout which could be considered a separate event triggered by the WaitFor function

Reply:
Resolved in https://github.com/smartdevicelink/sdl_core/pull/2912/commits/7d556a33dc13e3411e22c3e07eaad330c45a1e8b